File: python/Figure4_1.py
# Author: Kenji Kashima
# Date  : 2025/04/01
# Note  : pip install control

import numpy as np
import matplotlib.pyplot as plt
import config

np.random.seed(100)

a = 1.1;        # system matrix
alpha = -0.2;   # parameter in control law
x_max = 5;      # view range

def figure4_1a(k_bar:int=50):
    '''
        k_bar - total steps
    '''
    figsize = config.global_config(type=1)
    
    x = np.ones(k_bar+1)       # state with v=0
    for k in range(k_bar):
        u = alpha * (a + alpha) ** k  # control law
        x[k+1] = a * x[k] + u
    
    plt.figure(figsize=figsize)
    plt.plot(x,linewidth=1,color=[0,0,0])
    plt.xlim([0,k_bar])
    plt.ylim([-x_max,x_max])
    plt.xlabel(r'$k$')
    plt.ylabel(r'$x_k$')
    plt.tight_layout()
    plt.grid()
    plt.savefig("./Figure4_1a.pdf")
    plt.show()

def figure4_1b(k_bar:int=50,n_sample:int=20):
    '''
        k_bar - total steps
        n_sample - number of samples
    '''
    figsize = config.global_config(type=1)

    x = np.ones(k_bar+1)       # state with v=0
    x_v = np.ones([n_sample,k_bar+1]) # state with v~N(0,1)
    for k in range(k_bar):
        v = np.random.randn(n_sample) # n_sample independent standard normal distributions
        u = alpha * (a + alpha) ** k  # control law
        x[k+1] = a * x[k] + u
        x_v[:,k+1] = a * x_v[:,k] + u + 0.1 * v
    
    plt.figure(figsize=figsize)
    for i in range(n_sample):
        plt.plot(x_v[i,:],linewidth=0.2,color=[0.7,0.7,0.7])

    plt.plot(x,linewidth=1,color=[0,0,0])
    plt.xlim([0,k_bar])
    plt.ylim([-x_max,x_max])
    plt.xlabel(r'$k$')
    plt.ylabel(r'$x_k$')
    plt.tight_layout()
    plt.grid()
    plt.savefig("./Figure4_1b.pdf")
    plt.show()

if __name__ == '__main__':
    figure4_1a(k_bar=50)
    figure4_1b(k_bar=50,n_sample=20)

File: matlab/Figure4_1.m
% Author: Kenji Kashima
% Date  : 2025/09/01

clear;close all; rng(1); % random seed

k_bar = 50;       % total steps
n_sample = 20;  % number of samples
a = 1.1;        % system matrix
alpha = -0.2;   % parameter in control law
x_max = 5;      % view range


x = ones(1,k_bar+1);          % state with v=0
x_v = ones(n_sample,k_bar+1); % state with v~N(0,1)

for k = 1:k_bar
    v = randn(n_sample,1);  % n_sample independent standard normal distributions
    u = alpha*(a+alpha)^(k-1); % control law
    x(k+1) = a*x(k) + u;
    x_v(:,k+1) =a*x_v(:,k) + u + 0.1*v;
end

% Figure 4.1(a)
figure('Name','Figure 4.1(a)'); hold on; grid on;
plot(0:k_bar,x','LineWidth',1,'color',[0,0,0]+0); 
xlabel('$k$','Fontsize',16,'Interpreter','latex')
ylabel('$x_k$','Fontsize',16,'Interpreter','latex')
ylim([-x_max,x_max]);
movegui('northwest')

% Figure 4.1(b)
figure('Name','Figure 4.1(b)'); hold on; grid on;
plot(0:k_bar,x_v','LineWidth',0.2,'color',[0,0,0]+0.7); 
plot(0:k_bar,x','LineWidth',1,'color',[0,0,0]+0); 
xlabel('$k$','Fontsize',16,'Interpreter','latex')
ylabel('$x_k$','Fontsize',16,'Interpreter','latex')
ylim([-x_max,x_max]);
movegui('northeast')




